core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -63,13 +65,23 @@ public static <T> CloseableIterable<T> filter(CloseableIterable<T> rows, Functio
     return equalityFilter.filter(rows);
   }
 
-  public static <T> CloseableIterable<T> filter(CloseableIterable<T> rows, Function<T, Long> rowToPosition,
-                                                PositionDeleteIndex deleteSet) {
-    if (deleteSet.isEmpty()) {
-      return rows;
-    }
+  public static <T> CloseableIterable<T> markDeleted(CloseableIterable<T> rows, Predicate<T> isDeleted,
+                                                     Consumer<T> markDeleted) {
+    return CloseableIterable.transform(rows, row -> {
+      if (isDeleted.test(row)) {
+        markDeleted.accept(row);
+      }
+      return row;
+    });
+  }
 
-    PositionSetDeleteFilter<T> filter = new PositionSetDeleteFilter<>(rowToPosition, deleteSet);
+  public static <T> CloseableIterable<T> filter(CloseableIterable<T> rows, Predicate<T> shouldKeep) {
```

Comment:
I like the new filter method API. only concern is for compatibility. this is a public static method from a public class. Probably very few users call this API directly. if we want to be safe, we can keep the old API and mark it as deprecated.

Reply:
That's a good point. I had the same concern. However, we've discussed what should be considered as APIs in one of the community sync. we agreed that only consider the public things in the API module as APIs. cc @rdblue. With that, it should be OK to change this public method?